### PR TITLE
MPU Hal

### PR DESCRIPTION
**Pull request by DrZlo13**
# What's new

- MPU HAL
- NULL address is write and read protected now
- Thread stack is protected from overflow now

# Verification 

- idk

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Review by hedger — DISMISSED** [review]

File: firmware/targets/f7/furi_hal/furi_hal_interrupt.c [hunk 1]
```diff
@@ -241,6 +246,20 @@ void HardFault_Handler() {
 }
 
 void MemManage_Handler() {
+    if((READ_BIT(SCB->CFSR, SCB_CFSR_MMARVALID_Msk) >> SCB_CFSR_MMARVALID_Pos) == 1) {
```
Comment (hedger) [inline firmware/targets/f7/furi_hal/furi_hal_interrupt.c]: FURI_BIT() ?

File: firmware/targets/f7/furi_hal/furi_hal_interrupt.c [hunk 2]
```diff
@@ -241,6 +246,20 @@ void HardFault_Handler() {
...
+        uint32_t memfault_address = SCB->MMFAR;
+        if(memfault_address < (1024 * 1024)) {
+            // from 0x00 to 1MB, see FuriHalRegionNULL
+            furi_crash("NULL pointer dereference");
+        } else {
+            // write or read of MPU region 1 (FuriHalRegionStack)
+            furi_crash("MPU fault, possibly stack overflow");
+        }
+    } else if((READ_BIT(SCB->CFSR, SCB_CFSR_MSTKERR_Msk) >> SCB_CFSR_MSTKERR_Pos) == 1) {
```
Comment (hedger) [inline firmware/targets/f7/furi_hal/furi_hal_interrupt.c]: FURI_BIT() ?

File: firmware/targets/f7/furi_hal/furi_hal_mpu.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    furi_hal_mpu_disable();
+    LL_MPU_DisableRegion(region);
+    furi_hal_mpu_enable();
+}
+
+void furi_hal_mpu_set_stack_protection(uint32_t* stack) {
+    uint32_t stack_ptr = (uint32_t)stack;
+    uint32_t mask = ((1 << (FURI_HAL_MPU_STACK_PROTECT_REGION + 2)) - 1);
+    stack_ptr &= ~mask;
+    if(stack_ptr < (uint32_t)stack) stack_ptr += (mask + 1);
```
Comment (hedger) [inline firmware/targets/f7/furi_hal/furi_hal_mpu.c:62]: comment please about aligning protected area to size boundary

File: firmware/targets/furi_hal_include/furi_hal_mpu.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    FuriHalRegionSize16MB = 0x17U,
+    FuriHalRegionSize32MB = 0x18U,
+    FuriHalRegionSize64MB = 0x19U,
+    FuriHalRegionSize128MB = 0x1AU,
+    FuriHalRegionSize256MB = 0x1BU,
+    FuriHalRegionSize512MB = 0x1CU,
+    FuriHalRegionSize1GB = 0x1DU,
+    FuriHalRegionSize2GB = 0x1EU,
+    FuriHalRegionSize4GB = 0x1FU,
+} FuriHalRegionSize;
```
Comment (hedger) [inline firmware/targets/furi_hal_include/furi_hal_mpu.h]: FuriHalMpu...

File: firmware/targets/furi_hal_include/furi_hal_mpu.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+typedef enum {
+    FuriHalRegionNULL = 0x00, // region 0 used to protect null pointer dereference
+    FuriHalRegionStack = 0x01, // region 1 used to protect stack
+    FuriHalRegion2 = 0x02,
+    FuriHalRegion3 = 0x03,
+    FuriHalRegion4 = 0x04,
+    FuriHalRegion5 = 0x05,
+    FuriHalRegion6 = 0x06,
+    FuriHalRegion7 = 0x07,
+} FuriHalRegion;
```
Comment (hedger) [inline firmware/targets/furi_hal_include/furi_hal_mpu.h]: FuriHalMpu...

File: applications/nfc/nfc.c [hunk 6]
```diff
@@ -238,7 +238,7 @@ int32_t nfc_app(void* p) {
     char* args = p;
 
     // Check argument and run corresponding scene
-    if((*args != '\0')) {
+    if(args) {
```
Comment (hedger) [inline applications/nfc/nfc.c]: if args && strlen(args)

File: applications/lfrfid/lfrfid_app.cpp [hunk 7]
```diff
@@ -74,7 +74,7 @@ void LfRfidApp::run(void* _args) {
 
     make_app_folder();
 
-    if(strlen(args)) {
+    if(args) {
```
Comment (hedger) [inline applications/lfrfid/lfrfid_app.cpp]: if args && strlen(args)

**Review by hedger — CHANGES_REQUESTED** [review]

**Review by hedger — DISMISSED** [review]

**Review by skotopes — APPROVED** [review]

**Pull request merged**